pex/pex_bootstrapper.py

```diff
@@ -72,33 +72,40 @@ def _iter_interpreters():
         # type: () -> Iterator[InterpreterOrError]
         seen = set()
 
-        paths = None  # type: Optional[MutableSet[str]]
-        if path:
-            paths = OrderedSet(os.path.realpath(p) for p in path.split(os.pathsep))
+        normalized_paths = (
+            OrderedSet(os.path.realpath(p) for p in path.split(os.pathsep)) if path else None
+        )
 
+        # Check if the current interpreter is valid. If `path` was specified, the current
+        # interpreter must be included in the value; otherwise, we always yield the interpreter.
         current_interpreter = PythonInterpreter.get()
         if not _valid_path or _valid_path(current_interpreter.binary):
-            if paths:
-                # Prefer the current interpreter if present on the `path`.
+            if normalized_paths:
                 candidate_paths = frozenset(
                     (current_interpreter.binary, os.path.dirname(current_interpreter.binary))
                 )
-                candidate_paths_in_path = candidate_paths.intersection(paths)
+                candidate_paths_in_path = candidate_paths.intersection(normalized_paths)
                 if candidate_paths_in_path:
                     for p in candidate_paths_in_path:
-                        paths.remove(p)
+                        normalized_paths.remove(p)
                     seen.add(current_interpreter)
                     yield current_interpreter
             else:
-                # We may have been invoked with a specific interpreter, make sure our sys.executable is
-                # included as a candidate in this case.
                 seen.add(current_interpreter)
                 yield current_interpreter
 
-        for interp in PythonInterpreter.iter_candidates(paths=paths, path_filter=_valid_path):
-            if interp not in seen:
-                seen.add(interp)
-                yield interp
+        # NB: We only check for other interpreters if there are remaining candidate interpreters
+        # _or_ if the user left off `path`, meaning to default to looking at $PATH.
+        #
+        # There is a tricky edge case where `path` was specified, but its only entry is the current
+        # interpreter so it was already removed.
+        if normalized_paths or path is None:
```

Comment:
AFAICT this is the only actual logic change? The other changes (renaming of `paths` to `normalized_paths`) here makes it hard to see.

Reply:
I don't actually understand why this fixes the problem. Or, more precisely, why there was a problem in the first place. AFAICT we only remove the current interpreter if we're about to yield it. So it should be yielded by this function, and then I don't see why it matters what this `iter_candidates` does? 

If I can't figure out that edge case by reading the code and the comment then the comment needs elaborating and, ideally, the code needs simplifying...

Reply:
> So it should be yielded by this function, and then I don't see why it matters what this iter_candidates does?

We yield current interpreter, then remove it from the `normalized_paths`. Now, `normalized_paths=[]`. This means that `iter_candidates` will now see that it's empty, and will look like the user did not specify `--pex-path`, which triggers `iter_candidates` looking up the `$PATH`.

I'll revisit this all for readability.

Reply:
OK, that was extremely non-obvious, but makes sense.